util/java/libtiled-java/src/main/java/org/mapeditor/io/TMXMapReader.java

```diff
@@ -219,10 +235,15 @@ private TileSet unmarshalTilesetFile(InputStream in, URL file) throws Exception
             tsNode = tsNodeList.item(0);
             if (tsNode != null) {
                 set = unmarshalTileset(tsNode);
-                if (set.getSource() != null) {
-                    System.out.println("Recursive external tilesets are not supported.");
+
+                // if using cached tilesets - we should already have correct source
+                if (!settings.reuseCachedTilesets) {
+                    if (set.getSource() != null) {
```

Comment:
I think something is wrong with this check, even apart from it triggering a false warning in case a cached tileset is returned. If a recursive external tileset really happens, I think this code may never get reached and it would instead recurse indefinitely before.

Reply:
my bad, fixed it the way I think this should work